src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs

```diff
@@ -470,14 +467,48 @@ protected override Task<PackageSpec> GetPackageSpecAsync(CancellationToken ct)
             return GetPackageSpecAsync(NullSettings.Instance);
         }
 
-        protected override IEnumerable<PackageReference> FetchInstalledPackagesList(IEnumerable<LibraryDependency> libraries, NuGetFramework targetFramework, IList<LockFileTarget> targets)
+        protected override (IReadOnlyList<PackageReference>, FrameworkInstalledPackages) FetchInstalledPackagesList(IEnumerable<LibraryDependency> libraries, NuGetFramework targetFramework, IReadOnlyList<LockFileTarget> targets, Dictionary<string, ProjectInstalledPackage> installedPackages)
+        {
+            return GetPackageReferences(libraries, targetFramework, installedPackages, targets);
+        }
+
+        protected override (IReadOnlyList<PackageReference>, FrameworkInstalledPackages) FetchTransitivePackagesList(NuGetFramework targetFramework, IReadOnlyList<LockFileTarget> targets, Dictionary<string, ProjectInstalledPackage> installedPackages, Dictionary<string, ProjectInstalledPackage> transitivePackages)
+        {
+            return GetTransitivePackageReferences(targetFramework, installedPackages, transitivePackages, targets);
+        }
+
+        // To avoid race condition, we work on copy of cache.
+        protected override (Dictionary<string, ProjectInstalledPackage> installedPackagesCopy, Dictionary<string, ProjectInstalledPackage> transitivePackagesCopy) GetCacheCopy()
         {
-            return GetPackageReferences(libraries, targetFramework, InstalledPackages, targets);
+            return (new Dictionary<string, ProjectInstalledPackage>(InstalledPackages), new Dictionary<string, ProjectInstalledPackage>(TransitivePackages));
         }
 
-        protected override IReadOnlyList<PackageReference> FetchTransitivePackagesList(NuGetFramework targetFramework, IList<LockFileTarget> targets)
+        protected override void UpdatePackageListDetails(Dictionary<string, ProjectInstalledPackage> installedPackages, IEnumerable<FrameworkInstalledPackages> detectedInstalledPackageChanges)
         {
-            return GetTransitivePackageReferences(targetFramework, InstalledPackages, TransitivePackages, targets);
+            if (installedPackages == null || detectedInstalledPackageChanges == null)
+            {
+                return;
+            }
+
+            foreach (FrameworkInstalledPackages detectedInstalledPackageChange in detectedInstalledPackageChanges)
+            {
+                if (detectedInstalledPackageChange == null)
+                {
+                    continue;
+                }
+
```

Comment:
Below logic comes from PackageReferenceProject.cs: GetInstalledPackagesAsync >> FetchInstalledPackagesList >> GetPackageReferences >> GetPackageReferenceUtility.cs: UpdateResolvedVersion

Before we do mutation while method still running now apply diff after finish.
https://github.com/NuGet/NuGet.Client/blob/fc721e9dc7f52d482f680dccad8755f0e06450ad/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs#L46-L55